### PR TITLE
Monthly 2021 09+1

### LINKED ISSUE #251: Upgrade to Substrate's Latest Tag
Use `cargo update` with  `--precise 88c64e06471cc12aa9b25290f24d5566bcb5dd82` argument. to set substrate main to the correct Sha. it is based on master branch, and the sha we updated main and develop branch to

### LINKED ISSUE #256: Upgrade to Substrate's Latest Tag
For upgrading, we should consider to rebase on the integritee-master branch (which is already on substate v 4.0.0-dev)

Use cargo update with --precise 88c64e06471cc12aa9b25290f24d5566bcb5dd82 argument. to set substrate main to the correct Sha. it is based on master branch, and the sha we updated main and develop branch to

### PR DESCRIPTION
**Pull request by haerdib**
Updates to substrate septemeber + 1 tag and merges some updates from integritee master
Status of integritee master: 19 august , commit https://github.com/integritee-network/worker/commit/ed8628ef2d33f5aec94de413e19fc38c30d328fa

When reviewing this PR, focus on the polkadex code related changes -  this PR should not include any functionality changes in the polkadex specific code. Also, there are many file changes / additions that can be skipped over because they are not used by polkadex. Examples:
- `sidechain`
- `stf` (except the cli part)
- `top_pool`

Also, files that are newly created do not cause merge conflicts, so it is safe to assume that they do not introduce errors as they have been reviewed previously by integritee.

TODOs:

- [x] change all node dependencies to updated main / develop branches (pallet teerex, ocex, polkadex node, my_node_runtime..) (blocked by issue https://github.com/Polkadex-Substrate/Polkadex/issues/371)
- [x] compile code -> fix all errors that pop up along the way
- [x] rebase to current develop branch
- [x] merge [sgx-runtime](https://github.com/Polkadex-Substrate/sgx-runtime/pull/4) & change dependencies to point to master (blocked by task 2)
- [x] cargo fmt
- [x] Bigna: compare diff again -> any polkadex things changed?
- [x] fix clippy warnings
-> I removed all clippy warnings that are caused by this PR, but there are some warnings left which are already in the develop branch and/or newly popped up due to the addition of the "test" feature (some functions are only used in tests, even though they are not marked as such..)
I will create a separate issue for that -> https://github.com/Polkadex-Substrate/polkadexTEE-worker/issues/303
- [x] fix unit tests
- [x] test if worker & node can still talk with each other
- [ ] integration testing (@zktony ?)

For some files/constructs there has been a new concept introduced, which would require many changes in the code introduced by polkadex. Because of the hugeness of this PR I avoided complicated ingegration, and simply renamed the file to "polkadex_..." or added a FIXME comment. These files / functions should be integrated into the new construction. But in a separate PR so these changes will not be missed and can be reviewed properly.
This concerns the following files:
- worker/src/ipfs_polkadex.rs (new file: worker/src/ocall_bridge/ipfs_ocall.rs)
- worker/src/direct_incovation/polkadex_direct_server.rs (see files in same folder)
- alot of ocalls / ecalls should be moved to the ffi folder

I will create separate issues for these matters.

### CODE REVIEW

File: enclave/Cargo.toml [hunk 1]
```diff
@@ -16,173 +16,101 @@ crate-type = ["staticlib"]
...
+substratee-sidechain-primitives = { path = "../primitives/sidechain/primitives", default-features = false }
+substratee-stf = { path = "../stf", default-features = false, features = ["sgx"] }
+substratee-storage = { path = "../primitives/storage", default-features = false }
+substratee-worker-primitives= {path = "../primitives/worker", default-features = false }
+test-utils = { path = "../primitives/test-utils", default-features = false, features = ["sgx"] }
+
+# substrate deps
+frame-support = { tag = "monthly-2021-08", version = "4.0.0-dev", default-features = false, git = "https://github.com/paritytech/substrate.git" }
+sp-core = { tag = "monthly-2021-08", version = "4.0.0-dev", default-features = false, features = ["full_crypto"], git = "https://github.com/paritytech/substrate.git" }
+sp-finality-grandpa = { tag = "monthly-2021-08", version = "4.0.0-dev", default-features = false, git = "https://github.com/paritytech/substrate.git" }
```
Comment (simonsso) [inline enclave/Cargo.toml]: I think this should be branch=master and sha pointing to `monthly-2021-09+1`

File: primitives/api-client-extensions/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+[dependencies]
+codec = { package = "parity-scale-codec", version = "2.0.0", features = ["derive"] }
+
+# substrate
+sp-core = { tag = "monthly-2021-08", version = "4.0.0-dev",  git = "https://github.com/paritytech/substrate.git" }
+sp-finality-grandpa = { tag = "monthly-2021-08", version = "4.0.0-dev",  git = "https://github.com/paritytech/substrate.git" }
+sp-runtime = { tag = "monthly-2021-08", version = "4.0.0-dev",  git = "https://github.com/paritytech/substrate.git" }
+
+# polkadex
+substrate-api-client = { git = "https://github.com/Polkadex-Substrate/substrate-api-client", branch = "monthly-2021-08" }
```
Comment (simonsso) [inline primitives/api-client-extensions/Cargo.toml]: This link points to a repo where substrate is pointing to main, good

File: enclave/Cargo.toml (hunk 1 above)
Comment (haerdib) [reply enclave/Cargo.toml]: certainly. this will be changed, otherwise it will not compile

File: enclave/src/happy_path.rs [hunk 3]
```diff
@@ -16,8 +16,8 @@
 //     let gateway = create_mock_gateway();
 //     let alice: AccountId = get_account("happy_path_user_alice");
 //     let bob: AccountId = get_account("happy_path_user_bob");
-//     let token_a = AssetId::BTC;
-//     let token_b = AssetId::USD;
+//     let token_a = AssetId::Asset(2);
+//     let token_b = AssetId::Asset(3);
```
Comment (simonsso) [inline enclave/src/happy_path.rs]: These constants are defined using ISO standard for Fiat and 2 and 3 are reserved numbers @nuel77  have prepared this list for the frontend which should be imported if needed

```
COINCode_Bitcoin = 4294967297
COINCode_USD = 840
```

File: enclave/src/test_polkadex_gateway.rs [hunk 4]
```diff
@@ -297,7 +297,7 @@
 // //         user_uid: buy_order_user.clone(),
 // //         market_id: MarketId {
 // //             base: AssetId::POLKADEX,
-// //             quote: AssetId::DOT,
+// //             quote: AssetId::Asset(0),
```
Comment (simonsso) [inline enclave/src/test_polkadex_gateway.rs]: Why this large chunk of double commanded code?

Should be removed or made to work.

**Review by simonsso — CHANGES_REQUESTED** [review]

File: enclave/src/happy_path.rs (hunk 3 above)
Comment (haerdib) [reply enclave/src/happy_path.rs]: Cool. And where do I get this list?

File: enclave/src/test_polkadex_gateway.rs (hunk 4 above)
Comment (haerdib) [reply enclave/src/test_polkadex_gateway.rs]: I don't think I should make this PR any bigger with changes such as this (i.e. cleanups). This should be done in a separate PR, should it not?

File: enclave/src/happy_path.rs (hunk 3 above)
Comment (nuel77) [reply enclave/src/happy_path.rs]: @haerdib here is the list
https://github.com/Polkadex-Substrate/Polkadex-PolkaIDO-Frontend/blob/release-updated-currencies/generate-currency/cryptoList.json

File: worker/src/main.rs [hunk 5]
```diff
@@ -861,87 +728,53 @@ pub unsafe extern "C" fn ocall_send_release_extrinsic(
             vec![]
         }
     };
-    api.send_extrinsic(hex_encode(release_extrinsic_calls), XtStatus::Ready)
+    api.send_extrinsic(release_extrinsic_calls, XtStatus::Ready)
```
Comment (haerdib) [inline worker/src/main.rs:729]: in the new api, the extrinsic is hex_encoded in the `send_extrinsic` function

File: worker/src/main.rs [hunk 6]
```diff
@@ -861,87 +728,53 @@ pub unsafe extern "C" fn ocall_send_release_extrinsic(
...
         .unwrap();
     status
 }
 
-/// # Safety
-///
-/// FFI are always unsafe
-#[no_mangle]
-pub unsafe extern "C" fn ocall_send_block_and_confirmation(
-    confirmations: *const u8,
```
Comment (haerdib) [inline worker/src/main.rs:874]: Quite a git diff mess starting from here.
Effective change: `ocall_send_block_and_confirmation` has been moved to ocall_bridge/ffi, `get_blocks_to_sync` fn has been created (copy paste of the code above, that has been removed from [line 606-L637](https://github.com/Polkadex-Substrate/polkadexTEE-worker/pull/288/files?authenticity_token=w7eCHmslS9aK5g%2BUgs9Yj7rPTeprNyHXTABTmRTZhKu8n5VuyWB%2FU8XlOXVuKVGTq8s3azdnk%2B9Td%2FpH0VATUg%3D%3D&file-filters%5B%5D=.bin&file-filters%5B%5D=.der&file-filters%5B%5D=.pyc&file-filters%5B%5D=.rs&hide-deleted-files=true&w=1#diff-3381a66150dcfc521726bae16533dcb6463f5f799ba999a2f972bee40756a6bbL606-L637) )

File: client/src/main.rs [hunk 7]
```diff
@@ -426,29 +428,28 @@ fn main() {
...
-                    let (_to_encoded, to_encrypted) = match encode_encrypt(matches, to){
+                    let (_to_encoded, to_encrypted) = match encode_encrypt(matches, to) {
                         Ok((encoded, encrypted)) => (encoded, encrypted),
                         Err(e) => panic!("{}", e),
                     };
                     // compose the extrinsic
                     let xt: UncheckedExtrinsicV4<([u8; 2], Vec<u8>, u128, H256)> = compose_extrinsic!(
                         chain_api,
-                        "SubstrateeRegistry",
+                        TEEREX,
```
Comment (haerdib) [inline client/src/main.rs:444]: This line has not effectively changed -> const `TEEREX = "SubstrateeRegistry"`

**Review by simonsso — APPROVED** [review]
looks ok... I think

File: .github/workflows/build_and_test.unused.yml [hunk 8]
```diff
@@ -0,0 +1,222 @@
+# This file is copy pasted from Integritee - needs to be adapted first to be usable for polkadex
```
Comment (zktony) [inline .github/workflows/build_and_test.unused.yml:1]: Getting UnableToLoadPoniter Error

**Review by zktony — CHANGES_REQUESTED** [review]

**Review by zktony — APPROVED** [review]

**Pull request merged**